lib/worker/worker.go

```diff
@@ -0,0 +1,165 @@
+// Package worker provides interfaces to manage Linux processes
+package worker
+
+import (
+	"fmt"
+	"os/exec"
+	"sync"
+
+	"github.com/google/uuid"
+)
+
+// Worker is a Linux process manager
+// It keeps a registry of processes that have been requested
+// and exposes methods to manage those.
+// TODO: add logging to give some observability on worker's actions
+type Worker struct {
+	processRegistry map[ID]*ProcessHandle
+	mu              *sync.RWMutex
+}
+
+// ProcessRequest represents a request to execute a Linux process in the worker
+type ProcessRequest struct {
+	// Command is the command to be executed
+	Command string
+
+	// Args is a slice containing the arguments to the command
+	Args []string
+
+	// ResourceLimits specifies the resources that the process will have access to
+	// These translate to cgroup interface files configuration.
+	ResourceLimits ResourceLimits
+
+	// RequestedBy is the user that requested this process request
+	RequestedBy string
+}
+
+// NewWorker creates an instance of a Worker
+func NewWorker(cfg Config) *Worker {
+	registry := make(map[ID]*ProcessHandle)
+	return &Worker{
+		// resourceLimits: cfg.resourceLimitsDefault,
+		processRegistry: registry,
+		mu:              &sync.RWMutex{},
+	}
+}
+
+// ID is a unique ID that the worker uses to identify a process
+type ID string
+
+// ProcessHandle is a record of a process running in the worker
+type ProcessHandle struct {
+	process       Exec
+	outputHandler *OutputHandler
+}
+
+// ResourceLimits is a struct that holds requested resource limits
+// in a process request. These translate to cgroup interface files.
+type ResourceLimits struct {
+	MaxMemoryBytes int64
+	// ...
+}
+
+// StartProcess starts a new process and adds it to the worker
+// process registry. It does not wait for the process to terminate.
+func (w *Worker) StartProcess(req ProcessRequest) (ID, error) {
+	cmd := exec.Command(req.Command, req.Args...)
+
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return "", err
+	}
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return "", err
+	}
+
+	process := newProcess(req.Command, cmd, req.RequestedBy)
+
+	err = process.Start()
+	if err != nil {
+		return "", err
+	}
+
+	id := ID(uuid.NewString())
+
+	outputHandler, err := NewOutputHandler(stdout, stderr)
```

Comment:
Can you lose output between `process.Start()` and `NewOutputHandler()`. What if the process overflows its buffers and blocks?

Reply:
> Can you lose output between process.Start() and NewOutputHandler()

No. The cmd.StdoutPipe() and cmd.StdErrPipe() functions returns a `PipeReader`. This reader blocks until the PipeWriter writes to it or it is closed. Output remains until it is consumed.

Reply:
> What if the process overflows its buffers and blocks?
I am not sure I understood the question correctly but similarly to `PipeReader`, `PipeWriter` blocks until the new content is read.